net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,187 @@
+	if logf == nil {
+		logf = log.Printf
+	}
+
+	endpoints := []Endpoint{}
+
+	if derpMap != nil {
+		// If we have a DERP map, we can use the DERP hostnames and IPs as captive portal detection endpoints.
+		for _, region := range derpMap.Regions {
+			if region.Avoid {
+				continue
+			}
+			for _, node := range region.Nodes {
+				if node.IPv4 == "" {
+					continue
+				}
+				url := "http://" + node.IPv4 + "/generate_204"
+				p := DERPMapOther
+				if region.RegionID == preferredDERPRegionID {
+					p = DERPMapPreferred
+				}
+				e := Endpoint{url, http.StatusNoContent, "", true, p}
+				endpoints = append(endpoints, e)
+			}
+		}
+	} else {
+		logf("[v2] captivedetection: DERP map is nil, skipping DERP endpoints for captive portal detection")
+	}
+
+	// Let's also try the default Tailscale coordination server and admin console.
+	// These are likely to be blocked on some networks.
+	cs := Endpoint{"http://controlplane.tailscale.com/generate_204", http.StatusNoContent, "", false, Tailscale}
+	ac := Endpoint{"http://login.tailscale.com/generate_204", http.StatusNoContent, "", false, Tailscale}
+	endpoints = append(endpoints, cs, ac)
+
+	// Lastly, to be safe, let's also include some well-known captive portal detection URLs that are not under the
+	// tailscale.com umbrella. These are less likely to be blocked on public networks since blocking them
+	// would break captive portal detection for many devices.
+	switch goos {
+	case "windows":
+		endpoints = append(endpoints, Endpoint{"http://www.msftconnecttest.com/connecttest.txt", http.StatusOK, "Microsoft Connect Test", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.msftncsi.com/ncsi.txt", http.StatusOK, "Microsoft NCSI", false, Platform})
+	case "darwin", "ios":
+		endpoints = append(endpoints, Endpoint{"http://captive.apple.com/hotspot-detect.html", http.StatusOK, "Success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.thinkdifferent.us/", http.StatusOK, "Success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.airport.us/", http.StatusOK, "Success", false, Platform})
+	case "android":
+		endpoints = append(endpoints, Endpoint{"http://connectivitycheck.android.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://connectivitycheck.gstatic.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://play.googleapis.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://clients3.google.com/generate_204", http.StatusNoContent, "", false, Platform})
+	default:
+		endpoints = append(endpoints, Endpoint{"http://detectportal.firefox.com/success.txt", http.StatusOK, "success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://network-test.debian.org/nm", http.StatusOK, "NetworkManager is online", false, Platform})
```

Comment:
this could all probably be made less repetitive and remove all the "false" with a helper closure func that does the append